src/analyses/goto_check.cpp

```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
     return {};
 }
 
-/// Set a Boolean flag to a new value (via `set_flag`) and restore the previous
-/// value when the entire object goes out of scope.
+/// \brief Set a Boolean flag to a new value (via `set_flag`) and restore 
+/// the previous value when the entire object goes out of scope.
+///
+/// \remarks Calls to set_value are tracked to allow detecting doubles sets
+/// with different values and trigger an INVARIANT.
 class flag_resett
 {
 public:
-  /// Store the current value of \p flag and then set its value to \p new_value.
-  void set_flag(bool &flag, bool new_value)
+  /// \brief Store the current value of \p flag and
+  /// then set its value to \p new_value.
+  ///
+  /// \remarks an INVARIANT triggers iff the flag is set
+  /// more than once with different values.
+  void set_flag(
+    bool &flag,
+    bool new_value,
+    const irep_idt &flag_name,
+    const goto_programt::instructiont &instruction)
   {
+    INVARIANT(
+      !(seen(&flag) && flag != new_value),
+      "Flag " + id2string(flag_name) +
+        " set twice with incompatible values "
+        " at \n" +
+        instruction.source_location().pretty());
     if(flag != new_value)
     {
-      flags_to_reset.emplace_back(&flag, flag);
+      flags_to_reset.insert(std::pair<bool *, bool>(&flag, flag));
```

Comment:
I might be wrong, but I believe you can shorten this to `flags_to_reset.insert({&flag, flag});` with C++11.

Reply:
Or just `flags_to_reset.emplace(&flag, flag);` to avoid creating a temporary pair.